File: app/schemas/sku_forecast.py
"""Pydantic schemas for SKU-level forecast endpoints."""

from __future__ import annotations

from datetime import date
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class SkuForecastItem(BaseModel):
    product_id: int
    product_name: str
    product_type: str
    group_name: Optional[str] = None
    category_name: Optional[str] = None
    predicted_qty: float
    avg_price: Optional[float] = None
    estimated_revenue: Optional[float] = None


class SkuForecastBatchResponse(BaseModel):
    department_id: str
    department_name: str
    forecast_date: str
    items: List[SkuForecastItem]
    total_predicted_qty: float
    total_estimated_revenue: Optional[float] = None
    model_version: Optional[str] = None


class SkuForecastRangeResponse(BaseModel):
    department_id: str
    department_name: str
    from_date: str
    to_date: str
    daily_forecasts: List[SkuForecastBatchResponse]


class SkuTopNItem(BaseModel):
    product_id: int
    product_name: str
    product_type: str
    group_name: Optional[str] = None
    total_revenue: float
    total_qty: float
    avg_daily_qty: float
    rank: int


class SkuTopNResponse(BaseModel):
    department_id: Optional[str] = None
    period_days: int
    items: List[SkuTopNItem]


class SkuRetrainRequest(BaseModel):
    days: Optional[int] = Field(None, description="Training window in days (default: all available)")
    active_window_days: int = Field(30, description="SKU must have sold in this window to be 'active'")


class SkuRetrainResponse(BaseModel):
    status: str
    message: str
    metrics: Dict[str, Any]
    timestamp: str
    training_samples: int
    n_features: int
    n_unique_skus: int
    n_unique_departments: int


class SkuModelInfoResponse(BaseModel):
    status: str
    model_path: str
    n_features: int = 0
    training_metrics: Optional[Dict[str, Any]] = None
    trained_at: Optional[str] = None
    target_transform: Optional[str] = None


class SkuComparisonItem(BaseModel):
    date: str
    product_id: int
    product_name: str
    department_id: str
    department_name: str
    predicted_qty: Optional[float] = None
    actual_qty: float
    error: Optional[float] = None
    error_pct: Optional[float] = None


class SkuComparisonResponse(BaseModel):
    items: List[SkuComparisonItem]
    summary: Dict[str, Any]

File: frontend/src/types/sku-forecast.ts
export interface SkuForecastItem {
  product_id: number
  product_name: string
  product_type: string
  group_name: string | null
  category_name: string | null
  predicted_qty: number
  avg_price: number | null
  estimated_revenue: number | null
}

export interface SkuDailyForecast {
  department_id: string
  department_name: string
  forecast_date: string
  items: SkuForecastItem[]
  total_predicted_qty: number
  total_estimated_revenue: number | null
  model_version: string | null
}

export interface SkuForecastRange {
  department_id: string
  department_name: string
  from_date: string
  to_date: string
  daily_forecasts: SkuDailyForecast[]
}

export interface SkuTopNItem {
  product_id: number
  product_name: string
  product_type: string
  group_name: string | null
  total_revenue: number
  total_qty: number
  avg_daily_qty: number
  rank: number
}

export interface SkuTopNResponse {
  department_id: string | null
  period_days: number
  items: SkuTopNItem[]
}

export interface SkuModelInfo {
  status: string
  model_path: string
  n_features: number
  training_metrics: Record<string, number> | null
  trained_at: string | null
  target_transform: string | null
}

export interface SkuRetrainResponse {
  status: string
  message: string
  metrics: Record<string, unknown>
  timestamp: string
  training_samples: number
  n_features: number
  n_unique_skus: number
  n_unique_departments: number
}
